### PR TITLE
[REVIEW] unary.pyx: cython bindings

### LINKED ISSUE #3949: [FEA] Port unaryops.pyx to use non-legacy libcudf APIs

### PR DESCRIPTION
**Pull request by cwharris**
closes #3949

- [x] port `unary.pyx`
  - [x] `isnull` + `isna`
  - [x] `notnull` + `notna`
  - [x] `unary_operation`
- [x] port `transform.pyx`
  - [x] `transform`
  - [x] `nans_to_nulls`

### CODE REVIEW

**Comment by cwharris** [comment]
@kkraus14 @shwina 

- [ ] port `isnull` and `notnull`
- [ ] port `isna` and  `notna`
- [ ] port `nans_to_nulls`
- [ ] find and port the remaining unary operations

_*Discrepencies between `libcudf` and `libcudf++`:*_

libcudf++ offers `is_nan` support for floating point types. Should we use this for `column.isna`? If so, should we fail for non- floating point types, or fall back to `column.isnull`? Suggest we keep it as-is for now, so we don't have to write new tests during this port.

libcudf++ does not offer `nans_to_nulls`. could we use some combination of `is_nan` and `as_mask` to recreate this functionality? If so, is that the right approach?

**Comment by kkraus14** [comment]
> @kkraus14 @shwina
> 
> __Discrepencies between `libcudf` and `libcudf++`:__
> 
> libcudf++ offers `is_nan` support for floating point types. Should we use this for `column.isna`? If so, should we fail for non- floating point types, or fall back to `column.isnull`? Suggest we keep it as-is for now, so we don't have to write new tests during this port.

Agreed, `NaN` is a discrete value in cuDF whereas it's the missing value in Pandas. Lets keep null handling as is for now and if we need to add a function / adjust in the future we can.

> libcudf++ does not offer `nans_to_nulls`. could we use some combination of `is_nan` and `as_mask` to recreate this functionality? If so, is that the right approach?

Looks like it landed here: https://github.com/rapidsai/cudf/blob/branch-0.13/cpp/include/cudf/transform.hpp#L64

File: python/cudf/cudf/_libxx/unary.pyx [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+# def cast(Column input, object out_type):
+#     cdef column_view c_input = input.view()
+#     cdef type_id tid = np_to_cudf_types[np.dtype(out_type)]
+#     cdef data_type c_dtype = data_type(tid)
+
+#     with nogil:
+#         c_result = move(cpp_unary.cast(c_input, c_dtype))
+
+#     return Column.from_unique_ptr(move(c_result))
```
Comment (trxcllnt) [inline python/cudf/cudf/_libxx/unary.pyx]: @cwharris did you run into an issue calling the unary cast cpp method here? or just don't need it yet?

File: python/cudf/cudf/_libxx/includes/unary.pxd [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    cdef extern unique_ptr[column] unary_operation(
+        column_view input,
+        unary_op op)
+
+    cdef extern unique_ptr[column] is_null(column_view input)
+    cdef extern unique_ptr[column] is_valid(column_view input)
+    # cdef extern unique_ptr[column] cast(column_view input, data_type out_type)
+    cdef extern unique_ptr[column] is_nan(column_view input)
+    cdef extern unique_ptr[column] is_not_nan(column_view input)
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/includes/unary.pxd]:
```suggestion
    cdef extern unique_ptr[column] unary_operation(
        column_view input,
        unary_op op) except +

    cdef extern unique_ptr[column] is_null(column_view input) except +
    cdef extern unique_ptr[column] is_valid(column_view input) except +
    # cdef extern unique_ptr[column] cast(column_view input, data_type out_type) except +
    cdef extern unique_ptr[column] is_nan(column_view input) except +
    cdef extern unique_ptr[column] is_not_nan(column_view input) except +
```

File: python/cudf/cudf/_libxx/transform.pyx [hunk 3]
```diff
@@ -23,10 +23,22 @@ def bools_to_mask(Column col):
...
     buf = Buffer(rmm_db)
     return buf
+
+def nans_to_nulls(Column input):
+    cdef column_view c_input = input.view()
+    cdef pair[unique_ptr[device_buffer], size_type] c_output
+
+    with nogil:
+        c_output = move(libcudf_transform.nans_to_nulls(c_input))
+        c_buffer = move(c_output.first)
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/transform.pyx]: Do we need to declare this cdef?

File: python/cudf/cudf/_libxx/unary.pyx [hunk 4]
```diff
@@ -0,0 +1,100 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.
+
+from enum import IntEnum
+
+from cudf._libxx.column cimport Column
+from cudf._libxx.lib cimport *
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/unary.pyx]: explicitly import what you need

File: python/cudf/cudf/_libxx/unary.pyx [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+from cudf._libxx.lib import np_to_cudf_types
+from cudf._libxx.includes.unary cimport (
+    underlying_type_t_unary_op,
+    unary_op
+)
+
+cimport cudf._libxx.includes.unary as cpp_unary
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/unary.pyx]:
```suggestion
cimport cudf._libxx.includes.unary as libcudf_unary
```

File: python/cudf/cudf/_libxx/unary.pyx [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    BIT_INVERT = <underlying_type_t_unary_op> unary_op.BIT_INVERT
+    NOT = <underlying_type_t_unary_op> unary_op.NOT
+
+
+def unary_operation(Column input, object op):
+    cdef column_view c_input = input.view()
+    cdef unary_op c_op = <unary_op>(<underlying_type_t_unary_op> op)
+
+    with nogil:
+        c_result = move(
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/unary.pyx]: cdef this?

File: python/cudf/cudf/_libxx/unary.pyx [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        )
+
+    return Column.from_unique_ptr(move(c_result))
+
+
+def is_null(Column input):
+    cdef column_view c_input = input.view()
+
+    with nogil:
+        c_result = move(cpp_unary.is_null(c_input))
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/unary.pyx]: cdef this?

File: python/cudf/cudf/_libxx/unary.pyx [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+    return Column.from_unique_ptr(move(c_result))
+
+
+def is_valid(Column input):
+    cdef column_view c_input = input.view()
+
+    with nogil:
+        c_result = move(cpp_unary.is_valid(c_input))
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/unary.pyx]: cdef this?

File: python/cudf/cudf/_libxx/unary.pyx [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+#         c_result = move(cpp_unary.cast(c_input, c_dtype))
+
+#     return Column.from_unique_ptr(move(c_result))
+
+
+def is_nan(Column input):
+    cdef column_view c_input = input.view()
+
+    with nogil:
+        c_result = move(cpp_unary.is_nan(c_input))
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/unary.pyx]: cdef this?

File: python/cudf/cudf/_libxx/unary.pyx [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+    return Column.from_unique_ptr(move(c_result))
+
+
+def is_non_nan(Column input):
+    cdef column_view c_input = input.view()
+
+    with nogil:
+        c_result = move(cpp_unary.is_not_nan(c_input))
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/unary.pyx]: cdef this?

File: python/cudf/cudf/_libxx/unary.pyx (hunk 1 above)
Comment (cwharris) [reply python/cudf/cudf/_libxx/unary.pyx]: Was waiting for #4263, but it's since been closed.

File: python/cudf/cudf/_libxx/unary.pyx [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    column_view, mutable_column_view
+)
+from cudf._libxx.types import np_to_cudf_types
+from cudf._libxx.cpp.types cimport (
+    type_id,
+    data_type
+)
+from cudf._libxx.move cimport move
+from cudf._libxx.column import np_to_cudf_types, cudf_to_np_types
+from cudf._libxx.lib import np_to_cudf_types
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/unary.pyx]: Already imported this above
```suggestion
```

Comment (cwharris) [reply python/cudf/cudf/_libxx/unary.pyx]: weird. pretty sure this compiled and ran tests locally.

**Comment by cwharris** [comment]
@kkraus14 which plumbing are you refering to?

`Series` goes as far as `return self._unaryop("some_op")`, and `Index` doesn't refer to `unary` or `transform`.

**Comment by kkraus14** [comment]
> @kkraus14 which plumbing are you refering to?

The general unaryop plumbing for both Series / Index. IIRC Series is plumbed somewhat cleanly, but then Index just converts to a Series does the op, then converts back. All of that should move to `Frame` and then be inherited from Series / Index.

**Comment by cwharris** [comment]
Somehow or another I got distracted and forgot `unary.pyx` included more than just `unary_operation`. 

:man_facepalming:

I'll get it cleaned up.

**Comment by cwharris** [comment]
Having trouble figuring out why some tests are failing. For one set of tests, apparently a column is nullable when it shouldn't be (`test_rowwise_ops`, tests 0, 1, 2). I haven't determined how my code comes in to play with that yet.

**Comment by cwharris** [comment]
My `nans_to_nulls` was not accounting for some existing functionality. Existing `nans_to_nulls` in `unaryops.pyx` will return `None` if the result has no null values, making for an all-ones null mask, failing the test because the collumn is _nullable_.

Deciding where to fix this. Not sure if that logic should be kept in the `pyx`, since it's freeing the result of the operation, and if you wanted an all-ones null mask back, you'd have to generate it on the outside, resulting in double the work.

File: python/cudf/cudf/core/frame.py [hunk 12]
```diff
@@ -304,6 +304,44 @@ def _unaryop(self, op):
...
+        """
+        return self._is_valid()
+
+    def notna(self):
+        """Identify non-missing values. Alias for `notnull`.
+        """
+        return self._is_valid()
+
+    def _is_null(self):
+        result = self.copy()
```
Comment (kkraus14) [inline python/cudf/cudf/core/frame.py]: This is an unnecessary deep copy
```suggestion
        result = self.copy(deep=False)
```

File: python/cudf/cudf/core/frame.py [hunk 13]
```diff
@@ -304,6 +304,44 @@ def _unaryop(self, op):
...
+        for name, col in result._data.items():
+            result._data[name] = col.isnull()
+        return result
+        # return self.__class__._from_table(
+        #     libcudfxx.unary.is_null(self)
+        # )
+
+    def _is_valid(self):
+        result = self.copy()
```
Comment (kkraus14) [inline python/cudf/cudf/core/frame.py]:
```suggestion
        result = self.copy(deep=False)
```

File: python/cudf/cudf/core/frame.py [hunk 14]
```diff
@@ -304,6 +304,44 @@ def _unaryop(self, op):
...
+
+    def notna(self):
+        """Identify non-missing values. Alias for `notnull`.
+        """
+        return self._is_valid()
+
+    def _is_null(self):
+        result = self.copy()
+        for name, col in result._data.items():
+            result._data[name] = col.isnull()
```
Comment (kkraus14) [inline python/cudf/cudf/core/frame.py]: This feels a bit awkward to have to reach into the ColumnAccessor here. @shwina is this the expected way to handle this from `Frame`?

Comment (cwharris) [reply python/cudf/cudf/core/frame.py]: Yeah I’d rather have a libcudf api accepting a table_view, and use the established pattern, but thats not available in this case. Do we just assemble the table on this side, instead?

Comment (kkraus14) [reply python/cudf/cudf/core/frame.py]: I think what would be less awkward is constructing an `OrderedColumnDict`: https://github.com/rapidsai/cudf/blob/branch-0.13/python/cudf/cudf/utils/utils.py#L359 (can init the same as a normal `OrderedDict` but enforces types / lengths)

Then pass that into the `Table` init as `data` along with `self.index`: https://github.com/rapidsai/cudf/blob/branch-0.13/python/cudf/cudf/_libxx/table.pyx#L31

Maybe a little bit more verbose but feels more explicitly like the behavior you're trying to accomplish.

Comment (cwharris) [reply python/cudf/cudf/core/frame.py]: I'm pushing my changes so I can compare them better, but I still need to make changes, so feel free to ignore for now.

Comment (kkraus14) [reply python/cudf/cudf/core/frame.py]: Changes look good, resolving

File: python/cudf/cudf/core/frame.py [hunk 15]
```diff
@@ -299,10 +300,33 @@ def _copy_categories(self, other, include_index=True):
...
+        data = OrderedColumnDict(zip(self._column_names, data_columns))
+        return self.__class__._from_table(Frame(data, self._index))
+
+    def isna(self):
+        """Identify missing values. Alias for `isnull`
+        """
+        return self.isnull()
+
+    def notnull(self):
+        """Identify missing values.
```
Comment (cwharris) [inline python/cudf/cudf/core/frame.py]:
```suggestion
        """Identify non-missing values.
```

File: python/cudf/cudf/core/frame.py [hunk 16]
```diff
@@ -299,10 +300,33 @@ def _copy_categories(self, other, include_index=True):
         return self
 
     def _unaryop(self, op):
-        result = self.copy()
-        for name, col in result._data.items():
-            result._data[name] = col.unary_operator(op)
-        return result
+        data_columns = (col.unary_operator(op) for col in self._columns)
+        data = OrderedColumnDict(zip(self._column_names, data_columns))
```
Comment (shwina) [inline python/cudf/cudf/core/frame.py]: Technically `data` here can be a plain dict and the `Frame` constructor will handle converting it to the appropriate  mapping type (`ColumnAccessor`).

Comment (shwina) [reply python/cudf/cudf/core/frame.py]: Sorry - I realize I wasn't making a suggestion in the above comment: let's just use a plain `dict` here, and keep `OrderedColumnDict` an internal detail to `Frame`/`Table`.

**Pull request merged**